modules/face/src/face_alignment.cpp

```diff
@@ -0,0 +1,906 @@
+/**@brief structure determining split in regression tree
+*/
+struct splitr{
+        //!index1 Index of the first coordinates among the test coordinates for deciding split.
+        unsigned long index1;
+        //! index2 index of the second coordinate among the test coordinates for deciding split.
+        unsigned long index2;
+        //! thresh threshold for deciding the split.
+        float thresh;
+};
+/** @brief regression tree structure. Each leaf node is a vector storing residual shape.
+* The tree is represented as vector of leaves.
+*/
+struct regtree{
+    //! splits vector containing the splits made in the regression tree
+    vector<splitr> splits;
+    //! leaves represent leaf node of regression tree. Each leaf is a vector of points.
+    vector< vector<Point2f> > leaves;
+};
+/** @brief Represents a training sample
+*It contains current shape, difference between actual shape
+*and current shape. It also stores the image whose shape is being
+*detected.
+*/
+struct training_sample{
+    //! shapeResiduals vector which stores the residual shape remaining to be corrected.
+    vector<Point2f> shapeResiduals;
+    //! current_shape vector containing current estimate of the shape
+    vector<Point2f> current_shape;
+    //! actual_shape vector containing the actual shape of the face or the ground truth.
+    vector<Point2f> actual_shape;
+    //! image A mat object which stores the image.
+    Mat image ;
+    //! pixel_intensities vector containing pixel intensities of the coordinates chosen for testing
+    vector<float> pixel_intensities;
+};
+/** @brief This class is implementation of class FacemarkKazemi.
+*This class contains functions for training data and detection of
+*landmaks in images.
+*/
+class FacemarkKazemiimpl : public face::FacemarkKazemi{
+
+public:
+    virtual int getNumLandmarks() const {return numlandmarks;}
+    virtual int getNumFaces() const {return numfaces;}
+    /**@brief It assigns a meanshape from the precalculated mean shape
+    * if training data is not supplied.
+    */
+    FacemarkKazemiimpl(){
+        //get name  of file containing meanshape
+        ifstream f(meanshapefile.c_str());
+        string s;
+        //read value of annotations
+        while(getline(f,s)){
+            Point2f pt;
+            stringstream ss(s); // Turn the string into a stream.
+            string tok;
+            vector<string> coordinates;
+            while(getline(ss, tok,',')) {
+                coordinates.push_back(tok);
+            }
+            pt.x=(float)atof(coordinates[0].c_str());
+            pt.y=(float)atof(coordinates[1].c_str());
+            meanshape.push_back(pt);
+        }
+        //initialise other variables
+        numfaces =1;
+        numlandmarks =194;
+        //These variables are used for training data
+        //These are initialised as described in the research paper
+        //referenced above
+        cascade_depth = 10;
```

Comment:
These parameters should be supplyable to the code by the form of some sort of configuration file. I suggest using XML format for that. It allows user to play around, instead of getting this hardcoded.